src/app/qgsmaptooladdellipse.cpp

```diff
@@ -0,0 +1,96 @@
+/***************************************************************************
+    qgsmaptooladdellipse.cpp  -  map tool for adding ellipse
+    ---------------------
+    begin                : July 2017
+    copyright            : (C) 2017
+    email                : lbartoletti at tuxfamily dot org
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmaptooladdellipse.h"
+#include "qgscompoundcurve.h"
+#include "qgscurvepolygon.h"
+#include "qgsgeometryrubberband.h"
+#include "qgsgeometryutils.h"
+#include "qgslinestring.h"
+#include "qgsmapcanvas.h"
+#include "qgspoint.h"
+#include "qgisapp.h"
+
+QgsMapToolAddEllipse::QgsMapToolAddEllipse( QgsMapToolCapture *parentTool, QgsMapCanvas *canvas, CaptureMode mode )
+  : QgsMapToolCapture( canvas, QgisApp::instance()->cadDockWidget(), mode )
+  , mParentTool( parentTool )
+  , mTempRubberBand( nullptr )
+  , mEllipse( QgsEllipse() )
+{
+  clean();
+}
+
+QgsMapToolAddEllipse::~QgsMapToolAddEllipse()
+{
+  clean();
+}
+
+void QgsMapToolAddEllipse::keyPressEvent( QKeyEvent *e )
+{
+  if ( e && e->isAutoRepeat() )
+  {
+    return;
+  }
+
+  if ( e && e->key() == Qt::Key_Escape )
+  {
+    clean();
+    if ( mParentTool )
+      mParentTool->keyPressEvent( e );
+  }
+}
+
+void QgsMapToolAddEllipse::keyReleaseEvent( QKeyEvent *e )
+{
+  if ( e && e->isAutoRepeat() )
+  {
+    return;
+  }
+}
+
+void QgsMapToolAddEllipse::deactivate()
+{
+  if ( !mParentTool || mEllipse.isEmpty() )
+  {
+    return;
+  }
+
+  mParentTool->clearCurve();
+  mParentTool->addCurve( mEllipse.toLineString() );
+  clean();
+
+  QgsMapToolCapture::deactivate();
+}
+
+void QgsMapToolAddEllipse::activate()
+{
+  clean();
+  QgsMapToolCapture::activate();
+}
+
+void QgsMapToolAddEllipse::clean()
+{
+  if ( mTempRubberBand )
+  {
+    delete mTempRubberBand;
+    mTempRubberBand = nullptr;
```

Comment:
Do you think that `std::unique_ptr` could be used for `mTempRubberBand`?

Reply:
It wasn't not used in `QgsMapToolCircularString`.
I think that if we want use `std::unique_ptr` we have to make other changes in `QgsMapToolEdit` because we use `createGeometryRubberBand` which allow to have an alternativeBand like.
Maybe we have to change all other classes using rubberband. Another PR?